[core] FileNotFoundException for missing classes directory with analysis cache enabled

**Affects PMD Version:** 6.1.0

**Description:**

Original issue: https://issues.apache.org/jira/browse/MPMD-255

While trying to determine the checksum for the auxclasspath, we might encounter a `FileNotFoundException`, if one of the directories of the auxclasspath doesn't exist.

I think, we need to take care of this in PMD itself, to make it more safe.
But the m-pmd-p provides the auxclasspath and it obviously provides a "theoretical" classpath, even if the directory doesn't exist.
This might happen if the project really has no  `src/main/java` (e.g. just tests) or m-pmd-p is executed directly without compiling the project first.

Stacktrace:

```
Caused by: java.io.FileNotFoundException: my-project\target\classes (The system cannot find the file specified)
    at java.io.FileInputStream.open0 (Native Method)
    at java.io.FileInputStream.open (FileInputStream.java:195)
    at java.io.FileInputStream.<init> (FileInputStream.java:138)
    at java.io.FileInputStream.<init> (FileInputStream.java:93)
    at sun.net.www.protocol.file.FileURLConnection.connect (FileURLConnection.java:90)
    at sun.net.www.protocol.file.FileURLConnection.getInputStream (FileURLConnection.java:188)
    at java.net.URL.openStream (URL.java:1045)
    at net.sourceforge.pmd.cache.AbstractAnalysisCache.computeClassPathHash (AbstractAnalysisCache.java:182)
    at net.sourceforge.pmd.cache.AbstractAnalysisCache.checkValidity (AbstractAnalysisCache.java:113)
    at net.sourceforge.pmd.processor.AbstractPMDProcessor.processFiles (AbstractPMDProcessor.java:105)
    at net.sourceforge.pmd.PMD.processFiles (PMD.java:311)
    at org.apache.maven.plugins.pmd.PmdReport.processFilesWithPMD (PmdReport.java:503)
    at org.apache.maven.plugins.pmd.PmdReport.executePmd (PmdReport.java:432)
    at org.apache.maven.plugins.pmd.PmdReport.executePmdWithClassloader (PmdReport.java:331)
    at org.apache.maven.plugins.pmd.PmdReport.canGenerateReport (PmdReport.java:305)
```



**Running PMD through:** Maven

I also encountered this issue. I was trying to add incremental analysis to Gradle, but encountered this issue if I enabled incremental analysis and passed an `auxclasspath` containing an inexistent file:

```
Using auxclasspath: /Users/zhb/Projects/tmp/codeQuality/build/classes/java/main:/Users/zhb/Projects/tmp/codeQuality/build/classes/groovy/main:/Users/zhb/Projects/tmp/codeQuality/build/resources/main
```

What's worse, it's followed by a NPE:

```
Caused by: java.lang.NullPointerException
        at net.sourceforge.pmd.util.StringUtil.appendXmlEscaped(StringUtil.java:214)
        at net.sourceforge.pmd.renderers.XMLRenderer.end(XMLRenderer.java:132)
        at net.sourceforge.pmd.ant.Formatter.end(Formatter.java:117)
        at net.sourceforge.pmd.ant.internal.PMDTaskImpl.doTask(PMDTaskImpl.java:202)
        at net.sourceforge.pmd.ant.internal.PMDTaskImpl.execute(PMDTaskImpl.java:276)
        at net.sourceforge.pmd.ant.PMDTask.execute(PMDTask.java:49)
        at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:293)
        at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
        ... 60 more
```

Seems `pe.getFile()` is `null` in this line: https://github.com/pmd/pmd/blob/d25fbcf8c0b52b5e97dca3ea169f0eed5ca53ecd/pmd-core/src/main/java/net/sourceforge/pmd/renderers/XMLRenderer.java#L124

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -34,6 +34,8 @@ we have measured up to 10% improvements during Type Resolution, Symbol Table ana
 
 ### Fixed Issues
 
+*   all
+    *   [#988](https://github.com/pmd/pmd/issues/988): \[core] FileNotFoundException for missing classes directory with analysis cache enabled
 *   documentation
     *   [#994](https://github.com/pmd/pmd/issues/994): \[doc] Delete duplicate page contributing.md on the website
 *   java-bestpracrtices

--- a/pmd-core/src/main/java/net/sourceforge/pmd/cache/AbstractAnalysisCache.java
+++ b/pmd-core/src/main/java/net/sourceforge/pmd/cache/AbstractAnalysisCache.java
@@ -5,6 +5,7 @@
 package net.sourceforge.pmd.cache;
 
 import java.io.File;
+import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.net.URL;
 import java.net.URLClassLoader;
@@ -184,6 +185,8 @@ public abstract class AbstractAnalysisCache implements AnalysisCache {
                 while (IOUtils.skip(inputStream, Long.MAX_VALUE) == Long.MAX_VALUE) {
                     // just loop
                 }
+            } catch (final FileNotFoundException ignored) {
+                LOG.warning("Auxclasspath entry " + url.toString() + " doesn't exist, ignoring it");
             } catch (final IOException e) {
                 // Can this even happen?
                 LOG.log(Level.SEVERE, "Incremental analysis can't check auxclasspath contents", e);

--- a/pmd-core/src/test/java/net/sourceforge/pmd/cache/FileAnalysisCacheTest.java
+++ b/pmd-core/src/test/java/net/sourceforge/pmd/cache/FileAnalysisCacheTest.java
@@ -139,6 +139,21 @@ public class FileAnalysisCacheTest {
     }
     
     @Test
+    public void testAuxClasspathNonExistingAuxclasspathEntriesIgnored() throws MalformedURLException, IOException {
+        final RuleSets rs = mock(RuleSets.class);
+        final URLClassLoader cl = mock(URLClassLoader.class);
+        when(cl.getURLs()).thenReturn(new URL[] { new File(tempFolder.getRoot(), "non-existing-dir").toURI().toURL(), });
+        
+        setupCacheWithFiles(newCacheFile, rs, cl, sourceFile);
+        
+        final FileAnalysisCache analysisCache = new FileAnalysisCache(newCacheFile);
+        when(cl.getURLs()).thenReturn(new URL[] {});
+        analysisCache.checkValidity(rs, cl);
+        assertTrue("Cache believes unmodified file is not up to date after non-existing auxclasspath entry removed",
+                analysisCache.isUpToDate(sourceFile));
+    }
+    
+    @Test
     public void testAuxClasspathChangeWithoutDFAorTypeResolutionDoesNotInvalidatesCache() throws MalformedURLException, IOException {
         final RuleSets rs = mock(RuleSets.class);
         final URLClassLoader cl = mock(URLClassLoader.class);

